### PR TITLE
[`GPT2`] Add SDPA support

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
Adds torch's [SDPA](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) to the GPT2 model architecture (as another attention module). Possibly relevant https://github.com/huggingface/transformers/issues/28005.

I'll be adding some benchmarks like in #31031 for gpt2-large (same setup). Docs will be edited afterwards and I'll share my results in this thread. I've checked the two most important tests that validate sdpa vs eager: 
- `RUN_SLOW=True pytest tests/models/gpt2 -k "test_eager_matches_sdpa_generate" -s -vvvvv`
- `RUN_SLOW=True pytest tests/models/gpt2 -k "test_eager_matches_sdpa_inference" -s -vvvvv`

Both pass without any problems so it should probably be more about style / docs (after my benchmarks).  

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @younesbelkada @fxmarty @amyeroberts

### CODE REVIEW

**Comment by vasqu** [comment]
Oof sorry about the merge, should've squashed the commit :sweat:

**Comment by vasqu** [comment]
Training benchmark:
| Batch size | Seq len |  Time per batch (Eager - s) | Time per batch (SDPA - s) | Speedup (%) | Eager peak mem (MB) | SDPA peak mem (MB) |    Mem saving (%) |
|-----------:|--------:|----------------------------:|--------------------------:|------------:|--------------------:|-------------------:|------------------:|
|          1 |     128 |                       0.039 |                     0.032 |      23.042 |             3482.32 |            3494.62 |            -0.352 |
|          1 |     256 |                       0.073 |                     0.059 |       25.15 |             3546.66 |             3552.6 |            -0.167 |
|          1 |     512 |                       0.155 |                     0.118 |       30.96 |              4230.1 |            3665.59 |              15.4 |
|          1 |    1024 |                       0.316 |                     0.209 |      50.839 |             8682.26 |            4881.09 |            77.875 |
|          2 |     128 |                        0.07 |                      0.06 |      15.324 |              3557.8 |            3545.91 |             0.335 |
|          2 |     256 |                       0.143 |                     0.122 |       16.53 |              3901.5 |            3657.68 |             6.666 |
|          2 |     512 |                       0.267 |                     0.213 |      25.626 |             7062.21 |            4876.47 |            44.822 |
|          2 |    1024 |                         OOM |                     0.404 |           / |                 OOM |            8096.35 | SDPA does not OOM |
|          4 |     128 |                       0.134 |                     0.128 |       4.412 |             3675.79 |            3648.72 |             0.742 |
|          4 |     256 |                       0.243 |                     0.217 |      12.292 |             6129.76 |            4871.12 |            25.839 |
|          4 |     512 |                       0.494 |                     0.406 |      21.687 |             12466.6 |            8102.64 |            53.858 |
|          4 |    1024 |                         OOM |                     0.795 |           / |                 OOM |            14568.2 | SDPA does not OOM |

Inference benchmark:
| Batch size | Seq len | Per token latency Eager (ms) | Per token latency SDPA (ms) | Speedup (%) | Mem Eager (MB) | Mem SDPA (MB) | Mem saved (%) |
|-----------:|--------:|-----------------------------:|----------------------------:|------------:|---------------:|--------------:|--------------:|
|          1 |     128 |                        7.991 |                       6.968 |      14.681 |         1685.2 |       1701.32 |        -0.947 |
|          1 |     256 |                        8.462 |                       7.199 |      17.536 |        1745.49 |       1770.78 |        -1.428 |
|          1 |     512 |                         8.68 |                       7.853 |      10.529 |        1907.69 |       1921.29 |        -0.708 |
|          1 |     768 |                        9.101 |                       8.365 |       8.791 |        2032.93 |       2068.12 |        -1.701 |
|          2 |     128 |                        9.169 |                       9.001 |       1.861 |        1803.84 |        1811.4 |        -0.418 |
|          2 |     256 |                        9.907 |                        9.78 |       1.294 |        1907.72 |       1921.44 |        -0.714 |
|          2 |     512 |                       11.519 |                      11.644 |      -1.071 |        2176.86 |       2197.75 |        -0.951 |
|          2 |     768 |                       13.022 |                      13.407 |      -2.873 |         2464.3 |       2491.06 |        -1.074 |
|          4 |     128 |                       10.097 |                       9.831 |       2.709 |        1942.25 |       1985.13 |         -2.16 |
|          4 |     256 |                       11.599 |                      11.398 |       1.764 |        2177.28 |       2197.86 |        -0.937 |
|          4 |     512 |                       14.653 |                       14.45 |       1.411 |        2753.16 |       2772.57 |          -0.7 |
|          4 |     768 |                       17.846 |                      17.617 |       1.299 |        3327.04 |       3343.97 |        -0.506 |

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -1061,11 +1190,6 @@ def forward(
         # head_mask has shape n_layer x batch x n_heads x N x N
         head_mask = self.get_head_mask(head_mask, self.config.n_layer)
 
-        if inputs_embeds is None:
```
Comment (younesbelkada) [inline src/transformers/models/gpt2/modeling_gpt2.py:1064]: Why this has been removed ? This looks like a breaking change no?

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this contribution and for adding the extensive benchmark ! I left one open question, what do you think ?

File: src/transformers/models/gpt2/modeling_gpt2.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/gpt2/modeling_gpt2.py:1064]: Ah ic, yea it looks weird via the github diff but if you look above it's added right back there (L1137-1141). I just changed the order of operations to fit in with the sdpa mask creation.

**Review by younesbelkada — APPROVED** [review]
Thanks a lot !

**Review by ArthurZucker — COMMENTED** [review]
Left some small nits but overall good job! 👍🏻

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 2]
```diff
@@ -905,6 +1015,8 @@ def __init__(self, config):
         self.device_map = None
         self.gradient_checkpointing = False
         self._attn_implementation = config._attn_implementation
+        self._use_flash_attention_2 = self._attn_implementation == "flash_attention_2"
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt2/modeling_gpt2.py:1018]: In that case, I suppose `self._use_sdpa` should also be removed? Will change this with the according checks in the forward call.

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 3]
```diff
@@ -1022,11 +1134,24 @@ def forward(
...
+
         # Attention mask.
+        _use_sdpa = self._use_sdpa and output_attentions is False and head_mask is None
         if attention_mask is not None:
             attention_mask = attention_mask.view(batch_size, -1)
-            if self._attn_implementation == "flash_attention_2":
+            if self._use_flash_attention_2:
                 attention_mask = attention_mask if 0 in attention_mask else None
+            elif _use_sdpa:
+                attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt2/modeling_gpt2.py:1149]: Do you mean something like done in cohere? https://github.com/huggingface/transformers/blob/70c87138723b069634caf43f29ab754b5db8c50b/src/transformers/models/cohere/modeling_cohere.py#L951-L1029

Would need the same implementation here then. Seems conflicting with a few implementations to either reimplement `_update_causal_mask` or use the `_prepare_xx` methods.

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 4]
```diff
@@ -558,6 +561,115 @@ def _upad_input(self, query_layer, key_layer, value_layer, attention_mask, query
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Idea adapted from transformers.models.bert.modeling_bert.BertSdpaSelfAttention.__init__
+        # SDPA with memory-efficient backend is broken in torch==2.1.2 when using non-contiguous inputs and a custom
+        # attn_mask, so we need to call `.contiguous()`. This was fixed in torch==2.2.0.
+        # Reference: https://github.com/pytorch/pytorch/issues/112577
+        self.require_contiguous_qkv = version.parse(get_torch_version()) < version.parse("2.2.0")
+
+    # Adapted from GPT2Attention.forward and following other sdpa implementations
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt2/modeling_gpt2.py:580]: Will do :+1:

File: src/transformers/models/gpt2/modeling_gpt2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gpt2/modeling_gpt2.py]: thanks

File: src/transformers/models/gpt2/modeling_gpt2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gpt2/modeling_gpt2.py:1145]: yep no worries! 🤗

**Review by ArthurZucker — COMMENTED** [review]
Could you simply run the slow GPT2 tests to make sure everything is alright? As `SDPA` will be the new default!

**Comment by vasqu** [comment]
Did that commit 4811cb5580c4a5f166d357207ad00523c64d88c5 work or is it done locally only? @ArthurZucker

**Comment by vasqu** [comment]
Ok doesn't look like it ran in slow :sweat: i guess i need a different commit msg or perms?

**Comment by younesbelkada** [comment]
@vasqu if you have access to a GPU you can run them locally with `RUN_SLOW=1 pytest tests/models/gpt/test_modeling_gpt2.py`

**Comment by vasqu** [comment]
Ah ic, i will try and report back thx @younesbelkada; I don't have Flash Attention enabled tho (if that's an issue). Would require me to update my cuda which is a pain.

**Comment by younesbelkada** [comment]
That's shouldn't be an issue IMO, we just want to make sure the currently slow integration tests still pass with SDPA !

**Comment by vasqu** [comment]
[slow_test_gpt2_sample.txt](https://github.com/user-attachments/files/15741450/slow_test_gpt2_sample.txt)
[slow_test_gpt2_contrastive.txt](https://github.com/user-attachments/files/15741451/slow_test_gpt2_contrastive.txt)
[slow_test_gpt2.txt](https://github.com/user-attachments/files/15741453/slow_test_gpt2.txt)

When run all in one, three tests fail. Not sure why because when run in isolation only `test_gpt2_sample_max_time` fails. That could be fixed by alleviating the bounds at the last assert (because sdpa is faster).

**Comment by vasqu** [comment]
https://github.com/huggingface/transformers/blob/48d35b21789ad80a90ea242e46cb1d53e4db4f1c/tests/models/gpt2/test_modeling_gpt2.py#L832-L835

Should be changed with 
```python
        start = datetime.datetime.now()
        model.generate(input_ids, do_sample=False, max_time=None, max_length=256)
        duration = datetime.datetime.now() - start
        self.assertGreater(duration, datetime.timedelta(seconds=MAX_TIME))
        self.assertLess(duration, datetime.timedelta(seconds=1.5 * MAX_TIME))
```
That would follow the pattern of all assertions before that and then the tests also pass.

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 5]
```diff
@@ -832,7 +832,8 @@ def test_gpt2_sample_max_time(self):
         start = datetime.datetime.now()
         model.generate(input_ids, do_sample=False, max_time=None, max_length=256)
         duration = datetime.datetime.now() - start
-        self.assertGreater(duration, datetime.timedelta(seconds=1.5 * MAX_TIME))
+        self.assertGreater(duration, datetime.timedelta(seconds=MAX_TIME))
+        self.assertLess(duration, datetime.timedelta(seconds=1.5 * MAX_TIME))
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py:836]: Only failing test without this modification. I'm not sure if this is how it is intended. Maybe `assertGreater` should be `assertLess` instead?

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for working on this !

File: tests/models/gpt2/test_modeling_gpt2.py (hunk 5 above)
Comment (younesbelkada) [reply tests/models/gpt2/test_modeling_gpt2.py:836]: I think that this change is totally fine, now SDPA is used so the generation is faster

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 6]
```diff
@@ -1022,11 +1130,24 @@ def forward(
             position_ids = torch.arange(past_length, input_shape[-1] + past_length, dtype=torch.long, device=device)
             position_ids = position_ids.unsqueeze(0)
 
+        if inputs_embeds is None:
+            inputs_embeds = self.wte(input_ids)
+        position_embeds = self.wpe(position_ids)
+        hidden_states = inputs_embeds + position_embeds
+
         # Attention mask.
+        _use_sdpa = self._attn_implementation == "sdpa" and output_attentions is False and head_mask is None
```
Comment (ArthurZucker) [inline src/transformers/models/gpt2/modeling_gpt2.py:1139]: good!

**Review by ArthurZucker — APPROVED** [review]
Looks good to me! thanks for adding
